File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <bits/stdc++.h>

using namespace std;

bool one_diff(string a, string b){
    int n = a.size();
    int count = 0;
    for(int i = 0; i < n; i++){
        if(a[i] != b[i]) count++; //이거 언제 + 되더라?
        if(count > 1) return false;
    }
    return true;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    queue<pair<string, int>> q;
    vector<bool> visited(words.size(), false);
    
    // early return
    if(! count(words.begin(), words.end(), target)) return 0;
    
    // 로직 시작
    // q.push({begin, 0}); //이렇게 하면 자동으로 pair로 인식되나? vector일 수도 있고 tuple일 수도 있는데? <- 위에서 큐 정의할 때 해줘서 ㄱㅊ
    q.push({begin, 0});
    
    while(!q.empty()){
        auto a = q.front();
        string cur = a.first;
        int count = a.second;
        q.pop();
        
        for(int i = 0; i < words.size(); i++){
            string next = words[i];
            if(one_diff(cur, next) && !visited[i]){
                if(next == target) return count +1;
                q.push({next, count + 1});
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.kt
class Solution {
    
    fun one_different(a:String, b: String): Boolean{
        var count = 0
        for(i in 0 until a.length){
            if(a[i] != b[i]) count++
        }
        return (count == 1)
    }
    
    fun solution(begin: String, target: String, words: Array<String>): Int {
    
        val visited = BooleanArray(words.size)
        
        if (target !in words) return 0
        
        val q = ArrayDeque<Pair<String, Int>>()
        q.addLast(Pair(begin, 0))
        
        while(q.isNotEmpty()){
            val (cur, count) = q.removeFirst()
            
            if(cur == target) return count
            
            for(i in 0 until words.size){
                val word = words[i]
                if(one_different(word, cur) && !visited[i]){ // 근데 이렇게 하면 visited를 거를 수 있나? 여러개의 옵션 중에 하나만 선택하는 거 아닌가?
                    q.addLast(Pair(word, count + 1))
                    visited[i] = true
                }
            }
        }
        return 0
    }
}